### PR TITLE
Ignoring contextless lines during reporting

### PR DESCRIPTION
**Pull request by bbeattie-phxlabs**
As detailed in https://github.com/nedbat/coveragepy/issues/1350, coveragepy currently reports a single file with zero tests like

```
def foo_a(x, y):
  return x + y
```

as having 50% coverage. With [dynamic contexts](https://coverage.readthedocs.io/en/latest/contexts.html#dynamic-contexts) we can make use of line context to exclude lines executed outside of tests (imports, function definitions, etc) to arrive at an arguably more accurate lower bound of 0% for files without tests.

I suspect the PR will need to be augmented with additional documentation and tests to confirm this behaviour, but I figured getting the PR in for discussion might move things in the right direction.

### CODE REVIEW

File: coverage/results.py [hunk 1]
```diff
@@ -26,6 +26,13 @@ def __init__(self, data, precision, file_reporter, file_mapper):
         self.executed = executed
         self.missing = self.statements - self.executed
 
+        # Optionally ignore statements that have no context
```
Comment (bbeattie-phxlabs) [inline coverage/results.py:29]: This change and the change below are the crux of this PR.

**Comment by nedbat** [comment]
I'm still wondering whether this extra configuration is worth it.  I know you have an example where no tests results in 50% coverage, but in real projects the number will be much lower.  We could simply add an FAQ that explains that many lines get executed just by importing files.

**Comment by bbeattie-phxlabs** [comment]
To the thrust of my initial ticket that prompted the PR:

> My motivation is to coax developers in my org towards best practices without immediately requiring 100% code coverage. So a project template that says `--cov-fail-under=1` requires new projects to at least write one test, and we can go from there. But given coverage's current inclusion of let's call it "contextless code", I can't set a stable lower threshold at the moment.

I read on the overall size and complexity of the changes requested as fairly minimal, but I get that at the end of the day coveragepy is your baby. Happy to defer to your call on this. 👍

**Comment by bbeattie-phxlabs** [comment]
Alternatively, I'm just as happy to maintain my own fork with the `ignore_contextless` behaviour as a hardcoded default if there isn't sufficiently broad appeal to this option. :)

**Pull request closed without merging**